library/src/main/kotlin/com/freeletics/coredux/ReduxStore.kt

```diff
@@ -2,116 +2,121 @@ package com.freeletics.coredux
+ * @param stateReceiver function that receives new state updates
+ * @param sideEffects The sideEffects. See [SideEffect].
  * @param reducer The reducer.  See [Reducer].
  * @param S The type of the State
  * @param A The type of the Actions
+ *
+ * @return [ActionDispatcher] that should be use to dispatch new actions to [reduxStore]
  */
-@UseExperimental(ExperimentalCoroutinesApi::class, ObsoleteCoroutinesApi::class)
-fun <S: Any, A: Any> ReceiveChannel<A>.reduxStore(
+@UseExperimental(ExperimentalCoroutinesApi::class)
+fun <S: Any, A: Any> CoroutineScope.reduxStore(
     initialState: S,
-    sideEffects: List<SideEffect<S, A>>,
-    coroutineScope: CoroutineScope = GlobalScope,
+    stateReceiver: StateReceiver<S>,
+    sideEffects: List<SideEffect<S, A>> = emptyList(),
     reducer: Reducer<S, A>
-): ReceiveChannel<S> {
-    val output = Channel<S>()
-    val actionsChannel = BroadcastChannel<A>(1 + sideEffects.size)
-    var currentState = initialState
-
-    output.invokeOnClose {
-        this@reduxStore.cancel()
-        actionsChannel.close()
+): ActionDispatcher<A> {
+    val actionsReducerChannel = Channel<A>(Channel.UNLIMITED)
+    val actionsSideEffectsChannel = BroadcastChannel<A>(CONFLATED)
+
+    val actionDispatcher: ActionDispatcher<A> = { action ->
+        if (actionsReducerChannel.isClosedForSend) throw IllegalStateException("CoroutineScope is cancelled")
+
+        if (!actionsReducerChannel.offer(action)) {
+            throw IllegalStateException("Input actions overflow - buffer is full")
+        }
     }
 
-    // Sending initial state
-    coroutineScope.launch {
-        output.send(currentState)
+    coroutineContext[Job]?.invokeOnCompletion {
+        actionsReducerChannel.close(it)
+        actionsSideEffectsChannel.close(it)
     }
 
     // Starting reducer coroutine
-    coroutineScope.launch {
+    launch {
+        var currentState = initialState
+
+        // Sending initial state
+        stateReceiver(currentState)
+
+        // Starting side-effects coroutines
+        sideEffects.forEach { sideEffect ->
+            with (sideEffect) {
+                start(
+                    actionsSideEffectsChannel.openSubscription(),
+                    { currentState },
+                    actionsReducerChannel
+                )
+            }
+        }
+
         try {
-            for (action in actionsChannel.openSubscription()) {
-                try {
-                    currentState = reducer(currentState, action)
+            for (action in actionsReducerChannel) {
+                val newState = try {
+                    reducer(currentState, action)
                 } catch (e: Throwable) {
-                    output.close(ReducerException(currentState, action, e))
+                    throw ReducerException(currentState, action, e)
                 }
-                coroutineScope.launch {
-                    output.send(currentState)
+
+                if (newState != null) {
```

Comment:
Why does the `reducer` return null if the state does not change? Could it also return the same state again?

Reply:
It could return the same state. 

In RxRedux usually `.distinctUntilChanged()` is used to receive only updated states, but here it can't be used. I could introduce special `StateReceiver` wrapper or just use `null` to indicate that state is not changed, so I selected second solution, because it is easier for implementation and consuming sides :slightly_smiling_face:

Reply:
Actually, imho, came up with better extension function approach:
```kotlin
store.subscribe(::render.distinctUntilChanged())
```